tools/distribution/distrotest.py

```diff
@@ -0,0 +1,715 @@
+import logging
+import os
+import re
+import shutil
+from functools import cached_property
+from itertools import chain
+from typing import Callable, Iterable, Optional, Tuple, Type, Union
+
+import verboselogs
+
+import aiodocker
+
+from tools.base import checker, utils
+from tools.docker import utils as docker_utils
+
+DISTROTEST_CONFIG_PATH = "tools/distribution/distrotest.yaml"
+
+DOCKER_IMAGE_PREFIX = "envoybuild_"
+DOCKER_CONTAINER_PREFIX = "envoytest_"
+DOCKERFILE_TEMPLATE = """
+FROM {build_image}
+{env}
+
+ADD {install_dir} {install_mount_path}
+ADD {testfile_name} {test_mount_path}
+ADD {keyfile_name} {key_mount_path}
+RUN {build_command}
+
+CMD ["tail", "-f", "/dev/null"]
+"""
+
+
+class BuildError(Exception):
+    pass
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+class ContainerError(Exception):
+    pass
+
+
+class DistroTestConfig(object):
+    """Configuration object for distro tests
+
+    This holds the state/configuration across a batch of tests.
+
+    It also allows `DistroTest` to adapt `checker.Checker`.
+    """
+
+    packages_name = "packages"
+
+    def __init__(
+            self,
+            docker: aiodocker.Docker,
+            keyfile: str,
+            path: str,
+            tarball: str,
+            testfile: str,
+            maintainer: str,
+            version: str,
+            config_path: Optional[str] = None):
+        self.docker = docker
+        self._keyfile = keyfile
+        self.path = path
+        self.tarball = tarball
+        self._testfile = testfile
+        self.maintainer = maintainer
+        self.version = version
+        self._config_path = config_path
+
+    def __getitem__(self, k):
+        return self.config[k]
+
+    @cached_property
+    def config(self) -> dict:
+        """Configuration for test types - eg deb/rpm
+
+        This contains build information for different types of test image,
+        and some defaults for specific test configuration.
+        """
+        return utils.from_yaml(self.config_path)
```

Comment:
the idea crossed my mind, but tbh i didnt see the immediate benefit.

I guess though, as you say, the benefit is documentation and a schema to work with

there are 2 yaml configs 

this one defines the build/exec environment for the different container types - ie deb/rpm - its more complex but unlikely to change much and is not public-facing/configurable, at least with current implementation of the checker.

the other yaml schema (which comes as part of the next pr) is configurable/public-facing, but is super simple - basically just

```
debian_buster:
    image: debian:buster-slim
    ext: buster.changes
```

i would be happy following up to add protobuf schemas for these if you were ok with that

Reply:
the downside is i guess the dependency on protobuf

Reply:
Up to you, just thought it might have some win. My only strongly held opinion is that if we go with a schema, it should be protobuf, to reduce the proliferation of schema languages/tooling. Protobuf dependency isn't too bad I think, we have this in a fair bit of Python tooling already.

Reply:
yep, noted